### PR TITLE
iox-1803 Expand cmake configuration options

### LINKED ISSUE #1803: Increase configurability to allow smaller memory consumption
## Brief feature description

Just like the user can now configure the options `IOX_MAX_PUBLISHERS`, `IOX_MAX_SUBSCRIBER`, etc., it should be possible to configure all other hard-coded constants that significantly influence the memory footprint of the applications.

## Detailed information

Based on this discussion: https://github.com/eclipse-iceoryx/iceoryx/discussions/1770

### PR DESCRIPTION
**Pull request by JakubSosnovec**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
Based on this discussion: https://github.com/eclipse-iceoryx/iceoryx/discussions/1770

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1803

### CODE REVIEW

**Review by elfenpiff — COMMENTED** [review]
@JakubSosnovec I have no other words than: This is awesome!

Thank you so much for this great work we truly appreciate this!

**Review by elfenpiff — COMMENTED** [review]
You need to add the `small_memory` example as an exception to `tools/iceoryx_build_test.sh` (line 318 ff) since it does not contain any actual source code.

**Comment by JakubSosnovec** [comment]
@elfenpiff @elBoberido There are a few things to get your opinion on:
* Is the small_memory example a proper way how to demonstrate this? It isnt really an example since it doesnt have any source, it is rather a guide on how to achieve small memory footprint while keeping the basic helloworld demo functional. It could be another documentation guide. Not sure.
* I picked the variables that showed the most influence on memory usage. However, this leaves some constants still not configurable, because they dont influence memory so much. Now, it might sometimes be a bit strange, for example `MAX_RESPONSES_PROCESSED_SIMULTANEOUSLY` and `MAX_RESPONSE_QUEUE_CAPACITY` are configurable with cmake, but `MAX_REQUESTS_ALLOCATED_SIMULTANEOUSLY` is still hard-coded. Should we be consistent and make *all* the options configurable, or just keep it like this for now?

File: iceoryx_examples/small_memory/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+CMake option configuration for minimal memory usage. For example, as follows:
+
+```
+cmake -Bbuild -Hiceoryx_meta -Ciceoryx_examples/small_memory/options.cmake -DEXAMPLES=ON
+cmake --build build --target iox-roudi iox-cpp-publisher-helloworld iox-cpp-subscriber-helloworld
+```
+
+Please note that it might be impossible to build some other iceoryx examples,
+since the decreased maximum length of service and runtime name strings might
+cause some static assertion failures. This problem demonstrates that the user
```
Comment (budrus) [inline iceoryx_examples/small_memory/README.md:34]: @JakubSosnovec  These assertions are because the strings used in the other examples are too long then? Would it make sense to ensure all examples are compatible with the small_memory setup? Compiling maybe would be possible but we would get runtime problem as the resource limits are too small, I guess. So if we would like to ensure that all examples work, this would lead to bad compromises, e.g. having to increase the numbers again a bit because mempools are too small or not enough subscribers are possible. What do you think?

File: iceoryx_posh/BUILD.bazel [hunk 2]
```diff
@@ -25,18 +26,57 @@ configure_file(
...
-        "IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS": "256",
-        "IOX_MAX_CHUNKS_ALLOCATED_PER_PUBLISHER_SIMULTANEOUSLY": "8",
-        "IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY": "256",
-        "IOX_MAX_INTERFACE_NUMBER": "4",
-        "IOX_MAX_PUBLISHERS": "512",
-        "IOX_MAX_PUBLISHER_HISTORY": "16",
-        "IOX_MAX_SUBSCRIBERS": "1024",
-        "IOX_MAX_SUBSCRIBERS_PER_PUBLISHER": "256",
-    },
+        "@platforms//os:macos": {
```
Comment (JakubSosnovec) [inline iceoryx_posh/BUILD.bazel:31]: I am totally ignorant on Bazel, so I dont know how to do this. We just need to set the value of `IOX_MAX_RUNTIME_NAME_LENGTH` to 98 on Mac and on 100 on Linux, like in CMake. There must be a more elegant way, but this might just work.

**Comment by elBoberido** [comment]
@JakubSosnovec I can just repeat what elfenpiff said. Thanks very much for the PR. Unfortunately I did not yet have time to do a review but it's at the top of my todo list for next week :)

File: iceoryx_posh/cmake/IceoryxPoshDeployment.cmake [hunk 3]
```diff
@@ -27,46 +39,102 @@ if(NOT IOX_COMMUNICATION_POLICY)
...
-if(NOT IOX_MAX_PUBLISHER_HISTORY)
-    set(IOX_MAX_PUBLISHER_HISTORY 16)
-endif()
-message(STATUS "[i] IOX_MAX_PUBLISHER_HISTORY:" ${IOX_MAX_PUBLISHER_HISTORY})
-
-if(NOT IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY)
-    set(IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY 256)
-endif()
-message(STATUS "[i] IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY:" ${IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY})
+configure_option(
```
Comment (MatthiasKillat) [inline iceoryx_posh/cmake/IceoryxPoshDeployment.cmake:50]: OK, we keep the defaults from before here, right?

File: iceoryx_posh/BUILD.bazel [hunk 4]
```diff
@@ -25,18 +26,57 @@ configure_file(
...
+            "IOX_COMMUNICATION_POLICY": "ManyToManyPolicy",
+            "IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS": "256",
+            "IOX_MAX_CHUNKS_ALLOCATED_PER_PUBLISHER_SIMULTANEOUSLY": "8",
+            "IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY": "256",
+            "IOX_MAX_CLIENTS_PER_SERVER": "256",
+            "IOX_MAX_ID_STRING_LENGTH": "100",
+            "IOX_MAX_INTERFACE_NUMBER": "4",
+            "IOX_MAX_NODE_NAME_LENGTH": "100",
```
Comment (MatthiasKillat) [inline iceoryx_posh/BUILD.bazel:39]: OK, we add more fine grained control over some parameters (such as `IOX_MAX_NODE_NAME_LENGTH`).

File: iceoryx_posh/cmake/iceoryx_posh_deployment.hpp.in [hunk 5]
```diff
@@ -46,6 +47,21 @@ constexpr uint64_t IOX_MAX_PUBLISHER_HISTORY = static_cast<uint32_t>(@IOX_MAX_PU
 constexpr uint32_t IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY =
     static_cast<uint32_t>(@IOX_MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY@);
  constexpr uint32_t IOX_MAX_NUMBER_OF_NOTIFIERS = static_cast<uint32_t>(@IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS@);
+ constexpr uint32_t IOX_MAX_PROCESS_NUMBER = static_cast<uint32_t>(@IOX_MAX_PROCESS_NUMBER@);
```
Comment (MatthiasKillat) [inline iceoryx_posh/cmake/iceoryx_posh_deployment.hpp.in:50]: Correct me if I am wrong, but these appear all to be new options according to the diff. But I was sure we e.g. already had some of these for say request-response in sme way (e.g. `IOX_MAX_CLIENTS_PER_SERVER`).

Can you confirm that these are new parameters?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 6]
```diff
@@ -86,20 +86,20 @@ constexpr uint32_t MAX_GATEWAY_SERVICES = 2 * MAX_CHANNEL_NUMBER;
 // Client
 constexpr uint32_t MAX_CLIENTS = build::IOX_MAX_SUBSCRIBERS;
 constexpr uint32_t MAX_REQUESTS_ALLOCATED_SIMULTANEOUSLY = 4U;
-constexpr uint32_t MAX_RESPONSES_PROCESSED_SIMULTANEOUSLY = 16U;
-constexpr uint32_t MAX_RESPONSE_QUEUE_CAPACITY = 16U;
+constexpr uint32_t MAX_RESPONSES_PROCESSED_SIMULTANEOUSLY = build::IOX_MAX_RESPONSES_PROCESSED_SIMULTANEOUSLY;
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:89]: Ah yes, I can answer my own question now. We had the options but they were  hard-coded here.

However, it is somewhat inconsistent as some of them are still hard-coded. No need for changes, just an observation. We do not want some of them configurable, but the distinction may look arbitrary and not cleanly separated. Something for another issue, maybe.

**Review by MatthiasKillat — COMMENTED** [review]
Looks good and makes some previously non-configurable options configurable now. I did not try to build and run it yet, will do so on Monday. If we can solve the (potential) build issue with potential problems due to reduced string length we can merge it IMO.

@JakubSosnovec

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/convert.hpp [hunk 7]
```diff
@@ -81,6 +82,14 @@ class convert
     template <typename Destination>
     static bool fromString(const char* v, Destination& dest) noexcept;
 
+    /// @brief Sets dest from a given string. If the conversion fails false is
+    ///         returned and the value of dest is undefined.
+    /// @param[in] v string which contains the value of dest
+    /// @param[in] dest destination to which the value should be written
+    /// @return false = if the conversion fails otherwise true
+    template <uint64_t Capacity>
+    static bool fromString(const char* v, string<Capacity>& dest) noexcept;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/convert.hpp:91]: Is this tested? Maybe indirectly by passing a `string<Capacity>` to the previously existing template but then why do we need this overload in the first place?

If it is not tested we should add at least three tests, one with an empty string, one with a non-empty one and one with maximum length `Capacity`.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl [hunk 8]
```diff
@@ -86,15 +79,12 @@ inline bool convert::fromString<char>(const char* v, char& dest) noexcept
     return true;
 }
 
-/// @NOLINTJUSTIFICATION @todo iox-#260 convert must be refactored
-/// @NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
-template <>
-inline bool convert::fromString<string<100>>(const char* v, string<100>& dest) noexcept
+template <uint64_t Capacity>
+inline bool convert::fromString(const char* v, string<Capacity>& dest) noexcept
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl:83]: I see now why we would need to add this overload. In theory it should be tested in the way I requested with a `string<100>`.

File: iceoryx_examples/small_memory/README.md (hunk 1 above)
Comment (JakubSosnovec) [reply iceoryx_examples/small_memory/README.md:34]: @MatthiasKillat 

> Maybe it is worth considering to have a minimal with minimal string size that builds only one example?

Thats exactly what I did. The `small_memory` example that I added is basically just a guide on how to configure iceoryx just enough to be able to build the `icehello` example. I do know that other examples will fail, since some of them use very long strings.

File: iceoryx_posh/cmake/IceoryxPoshDeployment.cmake (hunk 3 above)
Comment (JakubSosnovec) [reply iceoryx_posh/cmake/IceoryxPoshDeployment.cmake:50]: Yes, these values are the defaults that were hard-coded before.

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 6 above)
Comment (JakubSosnovec) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:89]: Absolutely, I was also unsure about this inconsistency [here](https://github.com/eclipse-iceoryx/iceoryx/pull/1804#issuecomment-1327280593) . We can easily add at least the obvious ones, such as `MAX_REQUESTS_PROCESSED_SIMULTANEOUSLY`. I would make sense to me, I guess.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl (hunk 8 above)
Comment (JakubSosnovec) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl:83]: I added a simple test for `convert::fromString` overload. Also I actually modified the implementation to return `false` if the source size exceeds the `cxx::string` capacity.

File: iceoryx_posh/BUILD.bazel (hunk 2 above)
Comment (dkroenke) [reply iceoryx_posh/BUILD.bazel:31]: @JakubSosnovec I think we can leave it for now and live with the code duplication here, the issue seems to be that we cannot concatenate dicts with dicts coming from `select` statements properly.
Related issue: https://github.com/bazelbuild/bazel/issues/12457

File: iceoryx_posh/BUILD.bazel (hunk 4 above)
Comment (dkroenke) [reply iceoryx_posh/BUILD.bazel:39]: Shouldn't this be 98 here?

File: iceoryx_examples/small_memory/README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
+# icehello with minimal memory usage
```
Comment (dkroenke) [inline iceoryx_examples/small_memory/README.md:1]: Checked the example and it worked well thanks to the good description. Will continue reviewing.

Comment (dkroenke) [reply iceoryx_examples/small_memory/README.md:1]: @JakubSosnovec To make this example visible on the website you can add a simple redirection with a title in https://github.com/eclipse-iceoryx/iceoryx/tree/master/doc/website/examples to link to this markdown file.
Additionally please add an entry in https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/examples/.pages

Output looks like this: https://iceoryx.io/v2.0.2/examples/

Comment (mossmaurice) [reply iceoryx_examples/small_memory/README.md:1]: What to do if a new example is created is also documented [here](https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#how-to-add-a-new-example).

Comment (JakubSosnovec) [reply iceoryx_examples/small_memory/README.md:1]: @dkroenke @mossmaurice I have now added the example documentation and links according to https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#how-to-add-a-new-example, including the asciinema recording (`sed`ing the user to `hypnotoad` screwed up the recording a bit, because of the different lengths of my username and `hypnotoad`, so I used `htoad` instead).

Please take a look :)

**Comment by JakubSosnovec** [comment]
I have rebased to latest master, to fix some merge conflicts in copyright comment lines.

File: iceoryx_posh/cmake/IceoryxPoshDeployment.cmake [hunk 10]
```diff
@@ -27,46 +39,102 @@ if(NOT IOX_COMMUNICATION_POLICY)
...
 # note: don't change IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS value because it could break the C-Binding
-#if(NOT IOX_MAX_NUMBER_OF_NOTIFIERS)
+#configure_option(
+#    NAME IOX_MAX_NUMBER_OF_NOTIFIERS
+#    DEFAULT_VALUE 256
+#)
 set(IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS 256)
-#endif()
-#message(STATUS "[i] IOX_MAX_NUMBER_OF_NOTIFIERS:" ${IOX_MAX_NUMBER_OF_NOTIFIERS})
+
```
Comment (elBoberido) [inline iceoryx_posh/cmake/IceoryxPoshDeployment.cmake:137]: I think this is a remnant of the v2 release crunch time. AFAIK the note is not true any more and we could remove `INTERNAL`. @elfenpiff @dkroenke please correct me if I'm wrong

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl [hunk 11]
```diff
@@ -86,15 +79,12 @@ inline bool convert::fromString<char>(const char* v, char& dest) noexcept
     return true;
 }
 
-/// @NOLINTJUSTIFICATION @todo iox-#260 convert must be refactored
-/// @NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
-template <>
-inline bool convert::fromString<string<100>>(const char* v, string<100>& dest) noexcept
+template <uint64_t Capacity>
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/convert.inl:83]: \*cough\* `CString100` \*cough\*

**Review by elBoberido — APPROVED** [review]
Note to myself: Don't give estimates for PR reviews :sweat_smile: 

Looks good. Could be merged from my point of view.

**Review by elfenpiff — DISMISSED** [review]

**Comment by JakubSosnovec** [comment]
Rebased to latest master, to fix conflicts.

**Review by dkroenke — APPROVED** [review]
LGTM, thanks for the contribution.

**Review by elBoberido — APPROVED** [review]
